server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java

```diff
@@ -138,75 +137,23 @@ ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
-                    if (scoreDoc.shardIndex == request.shardRequestIndex()) {
-                        scoreDocs.add(scoreDoc);
-                    }
+            for (ScoreDoc scoreDoc : dfsKnnResults.scoreDocs()) {
+                if (scoreDoc.shardIndex == request.shardRequestIndex()) {
+                    scoreDocs.add(scoreDoc);
                 }
             }
             scoreDocs.sort(Comparator.comparingInt(scoreDoc -> scoreDoc.doc));
-            // It is possible that the different results refer to the same doc.
-            for (int i = 0; i < scoreDocs.size() - 1; i++) {
-                ScoreDoc scoreDoc = scoreDocs.get(i);
-                int j = i + 1;
-                for (; j < scoreDocs.size(); j++) {
-                    ScoreDoc otherScoreDoc = scoreDocs.get(j);
-                    if (otherScoreDoc.doc != scoreDoc.doc) {
-                        break;
-                    }
-                    scoreDoc.score += otherScoreDoc.score;
-                }
-                if (j > i + 1) {
-                    scoreDocs.subList(i + 1, j).clear();
-                }
-            }
-
             KnnScoreDocQueryBuilder knnQuery = new KnnScoreDocQueryBuilder(scoreDocs.toArray(new ScoreDoc[0]));
-            SearchSourceBuilder newSource = source.shallowCopy().knnSearch(List.of());
-            if (source.query() == null) {
-                newSource.query(knnQuery);
-            } else {
-                newSource.query(new BoolQueryBuilder().should(knnQuery).should(source.query()));
-            }
-            request.source(newSource);
-        } else {
-            // this path will keep knn queries separate for ranking per shard
-            // if there are multiple knn queries
-
-            List<QueryBuilder> rankQueryBuilders = new ArrayList<>();
-            if (source.query() != null) {
-                rankQueryBuilders.add(source.query());
-            }
-
-            for (DfsKnnResults dfsKnnResults : knnResults) {
-                List<ScoreDoc> scoreDocs = new ArrayList<>();
-                for (ScoreDoc scoreDoc : dfsKnnResults.scoreDocs()) {
-                    if (scoreDoc.shardIndex == request.shardRequestIndex()) {
-                        scoreDocs.add(scoreDoc);
-                    }
-                }
-                scoreDocs.sort(Comparator.comparingInt(scoreDoc -> scoreDoc.doc));
-                KnnScoreDocQueryBuilder knnQuery = new KnnScoreDocQueryBuilder(scoreDocs.toArray(new ScoreDoc[0]));
-                rankQueryBuilders.add(knnQuery);
-            }
-
-            BoolQueryBuilder searchQuery = new BoolQueryBuilder();
-            for (QueryBuilder queryBuilder : rankQueryBuilders) {
-                searchQuery.should(queryBuilder);
-            }
-
-            SearchSourceBuilder newSource = source.shallowCopy().query(searchQuery).knnSearch(List.of());
-            request.source(newSource);
-            request.rankQueryBuilders(rankQueryBuilders);
```

Comment:
All of this logic goes away since we build the boolean query on the shard when necessary as part of parsing the `SearchSourceBuilder` into a `DefaultSearchContext`. I admit I was surprised by how much code we were able to pull from the coordinator phases. Best part is it naturally works for dfs, query, and fetch :)